index.js

```diff
@@ -502,6 +506,455 @@
+      [TypeRep($.TypeVariable('f')), a, f(a)],
+      Z.of);
+
+  //# chain :: Chain m => (a -> m b) -> m a -> m b
+  //.
+  //. Curried version of [`Z.chain`][].
+  //.
+  //. ```javascript
+  //. > S.chain(x => [x, x], [1, 2, 3])
+  //. [1, 1, 2, 2, 3, 3]
+  //.
+  //. > S.chain(n => s => s.slice(0, n), s => Math.ceil(s.length / 2))('slice')
+  //. 'sli'
+  //.
+  //. > S.chain(S.parseInt(10), S.Just('123'))
+  //. Just(123)
+  //.
+  //. > S.chain(S.parseInt(10), S.Just('XXX'))
+  //. Nothing
+  //. ```
+  S.chain = def('chain', {m: [Z.Chain]}, [Fn(a, m(b)), m(a), m(b)], Z.chain);
+
+  //# join :: Chain m => m (m a) -> m a
+  //.
+  //. [Type-safe][sanctuary-def] version of [`Z.join`][].
+  //. Removes one level of nesting from a nested monadic structure.
+  //.
+  //. ```javascript
+  //. > S.join([[1], [2], [3]])
+  //. [1, 2, 3]
+  //.
+  //. > S.join([[[1, 2, 3]]])
+  //. [[1, 2, 3]]
+  //.
+  //. > S.join(S.Just(S.Just(1)))
+  //. S.Just(1)
+  //. ```
+  S.join = def('join', {m: [Z.Chain]}, [m(m(a)), m(a)], Z.join);
+
+  //# chainRec :: ChainRec m => TypeRep m -> (a -> Either (m b) (m b)) -> a -> m b
+  //.
+  //. Performs a [`chain`](#chain)-like computation with constant stack usage.
+  //. Similar to [`Z.chainRec`][], but curried and more convenient due to the
+  //. use of the Either type to indicate completion.
+  //.
+  //. ```javascript
+  //. > S.chainRec(Array,
+  //. .            s => s.length === 2 ? S.Right([s + '!', s + '?'])
+  //. .                                : S.Left([s + 'o', s + 'n']),
+  //. .            '')
+  //. ['oo!', 'oo?', 'on!', 'on?', 'no!', 'no?', 'nn!', 'nn?']
+  //. ```
+  S.chainRec =
+  def('chainRec',
+      {m: [Z.ChainRec]},
+      [TypeRep($.TypeVariable('m')), Fn(a, $Either(m(b), m(b))), a, m(b)],
+      function(typeRep, f, x) {
+        function step(next, done, x) {
+          var either = f(x);
+          return Z.map(either.isLeft ? next : done, either.value);
+        }
+        return Z.chainRec(typeRep, step, x);
+      });
+
+  //# extend :: Extend w => (w a -> b) -> w a -> w b
+  //.
+  //. Curried version of [`Z.extend`][].
+  //.
+  //. ```javascript
+  //. > S.extend(xs => xs.length, ['foo', 'bar', 'baz', 'quux'])
+  //. [4]
+  //. ```
+  S.extend =
+  def('extend', {w: [Z.Extend]}, [Fn(w(a), a), w(a), w(a)], Z.extend);
+
+  //# extract :: Comonad w => w a -> a
+  //.
+  //. [Type-safe][sanctuary-def] version of [`Z.extract`][].
+  S.extract =
+  def('extract', {w: [Z.Comonad]}, [w(a), a], Z.extract);
+
+  //# filter :: (Applicative f, Foldable f, Monoid (f a)) => (a -> Boolean) -> f a -> f a
+  //.
+  //. Curried version of [`Z.filter`][].
+  //.
+  //. See also [`filterM`](#filterM).
+  //.
+  //. ```javascript
+  //. > S.filter(S.odd, [1, 2, 3, 4, 5])
+  //. [1, 3, 5]
+  //.
+  //. > S.filter(S.odd, S.Just(9))
+  //. Just(9)
+  //.
+  //. > S.filter(S.odd, S.Just(4))
+  //. Nothing
+  //. ```
+  S.filter =
+  def('filter',
+      {f: [Z.Applicative, Z.Foldable, Z.Monoid]},
```

Comment:
This is a limitation of sanctuary-def. I'd love to work out how to properly capture this constraint.